src/Source.jl

```diff
@@ -240,9 +305,52 @@ Data.accesspattern(::Type{<:CSV.Source}) = Data.Sequential
 @inline Data.isdone(io::CSV.Source, row, col, rows, cols) = eof(io.io) || (!ismissing(rows) && row > rows)
 @inline Data.isdone(io::Source, row, col) = Data.isdone(io, row, col, size(io.schema)...)
 Data.streamtype(::Type{<:CSV.Source}, ::Type{Data.Field}) = true
-@inline Data.streamfrom(source::CSV.Source, ::Type{Data.Field}, ::Type{T}, row, col::Int) where {T} = CSV.parsefield(source.io, T, source.options, row, col)
 Data.reference(source::CSV.Source) = source.io.data
 
+struct Error <: Exception
+    result::Parsers.Result
+    row::Int
+    col::Int
+end
+
+@inline function parsefield(source::CSV.Source{P, I, DF, D}, ::Type{T}, row, col; kwargs...) where {P, I, DF, D, T}
+    D === Vector{Int} && Parsers.fastseek!(source.io, source.datapos[col])
+    r = Parsers.parse(source.parsinglayers, source.io, T; kwargs...)
+    D === Vector{Int} && setindex!(source.datapos, position(source.io), col)
+    if Parsers.ok(r.code)
+        return r.result
+    else
+        throw(Error(r, row, col))
+    end
+end
+
+@inline Data.streamfrom(source::Source, ::Type{Data.Field}, T, row, col::Int) = parsefield(source, Base.nonmissingtype(T), row, col)
+@inline Data.streamfrom(source::Source, ::Type{Data.Field}, ::Union{Type{Bool}, Union{Bool, Missing}}, row, col::Int) = parsefield(source, Bool, row, col; bools=source.bools)
+@inline Data.streamfrom(source::Source, ::Type{Data.Field}, ::Type{T}, row, col::Int) where {T <: Union{AbstractFloat, Missing}} = parsefield(source, Base.nonmissingtype(T), row, col; decimal=source.decimal)
+@inline Data.streamfrom(source::Source, ::Type{Data.Field}, ::Type{T}, row, col::Int) where {T <: Union{Dates.TimeType, Missing}} = parsefield(source, Base.nonmissingtype(T), row, col; dateformat=source.dateformat)
+@inline Data.streamfrom(source::Source, ::Type{Data.Field}, ::Type{Missing}, row, col::Int) = parsefield(source, Missing, row, col)
+
+@inline function get🐱(pool::CategoricalPool, val::Tuple{Ptr{UInt8}, Int})
```

Comment:
I don't really understand this trick. How is it possible that looking up `val` in `pool.invindex` works, given that `val` is a `Tuple` but the pool appears to contain `CategoricalString`s? Also `get!(pool, val)` should do the same thing.

Reply:
Yeah, it's a little tricksy. In Parsers.jl, we use `Tuple{Ptr{UInt8}, Int}` as a sort of poor man's WeakRefString; i.e. it's never returned to users directly, but used in a few places internally for performance. In particular, all strings are actually interned automatically using this Tuple{Ptr{UInt8}, Int} type and defining hash on it.

In this case, the `pool.invindex` has the actual strings values, so our Tuple type will hash to equal on the right level. The whole idea here is we want to avoid materializing the String in the first place, so we need to be able to compare a pointer & len w/ already materialized strings to see if we can just return a reference or if we need to materialize the string and add it to the pool.